File: 3367.maximize-sum-of-weights-after-edge-removals.py
#
# @lc app=leetcode id=3367 lang=python3
#
# [3367] Maximize Sum of Weights after Edge Removals
#

# @lc code=start
class Solution:
    def maximizeSumOfWeights(self, edges: List[List[int]], k: int) -> int:
        from collections import defaultdict
        # Create graph with node connections count
        node_connections = defaultdict(int)
        # Sort edges by weights in descending order to maximize sum when added
        edges.sort(key=lambda x: x[2], reverse=True)
        # Initialize max sum of weights
        max_weight_sum = 0
        for u, v, w in edges:
            if node_connections[u] < k and node_connections[v] < k:
                # Add edge if both nodes have less than k connections
                max_weight_sum += w
                node_connections[u] += 1
                node_connections[v] += 1
        return max_weight_sum
# @lc code=end

File: 3367.maximize-sum-of-weights-after-edge-removals.java
#
# @lc app=leetcode id=3367 lang=java
#
# [3367] Maximize Sum of Weights after Edge Removals
#

# @lc code=start
import java.util.*;
class Solution {
    public long maximizeSumOfWeights(int[][] edges, int k) {
        int n = edges.length + 1;
        List<int[]>[] tree = new ArrayList[n];
        for(int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for(int[] e : edges) {
            tree[e[0]].add(new int[]{e[1], e[2]});
            tree[e[1]].add(new int[]{e[0], e[2]});
        }
        return dfs(0, -1, tree, k)[0];
    }
    private long[] dfs(int u, int parent, List<int[]>[] tree, int k) {
        // dp[0]: max sum in subtree rooted at u
        // dp[1]: list of edge weights we can offer to parent for possible selection
        PriorityQueue<Long> gains = new PriorityQueue<>();
        long sum = 0;
        for(int[] nei : tree[u]) {
            int v = nei[0], w = nei[1];
            if(v == parent) continue;
            long[] child = dfs(v, u, tree, k);
            // child[0]: max sum in subtree rooted at v
            // child[1]: best gain if we keep edge u-v
            sum += child[0];
            gains.offer(child[1] + w);
            if(gains.size() > k) gains.poll();
        }
        long keep = 0;
        List<Long> list = new ArrayList<>();
        while(!gains.isEmpty()) {
            long val = gains.poll();
            keep += val;
            list.add(val);
        }
        return new long[]{sum + keep, list.size() == 0 ? 0 : Collections.max(list)};
    }
}
# @lc code=end